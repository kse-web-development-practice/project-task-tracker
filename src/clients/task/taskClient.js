async function addTask(name, description, deadline, importance, isCompleted, projectId) {
  const task = {
    name,
    description,
    deadline,
    importance,
    isCompleted,
    projectId
  }

  const response = await fetch(`${process.env.API_BASE_URL}/task/add`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(task)
  })

  if (response.status === 500) {
    throw new Error(`Error: ${response.statusText}`)
  }

  if (!response.ok) {
    return null
  }

  const body = await response.json()
  return body.task
}

async function getTasksByProjectId(projectId) {
  const response = await fetch(`${process.env.API_BASE_URL}/task/${projectId}`, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json'
    }
  })

  if (response.status === 500) {
    throw new Error(`Error: ${response.statusText}`)
  }

  if (!response.ok) {
    return null
  }

  const body = await response.json()
  return body.tasks
}

async function updateTask(id, updates) {
  const response = await fetch(`${process.env.API_BASE_URL}/task/update/${id}`, {
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(updates)
  })

  if (response.status === 500) {
    throw new Error(`Error: ${response.statusText}`)
  }

  if (!response.ok) {
    return null
  }

  const body = await response.json()
  return body
}

export default {
  addTask,
  getTasksByProjectId,
  updateTask
}
