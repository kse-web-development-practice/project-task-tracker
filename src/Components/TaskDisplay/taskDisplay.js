import React, { useState } from 'react'
import PropTypes from 'prop-types'
import styles from './taskDisplay.module.css'
import { ListCheckbox } from '../ListCheckbox/listCheckbox'
import { ListSelect } from '../ListSelect/listSelect'
import { ListCell } from '../ListCell/listCell'
import taskClient from '../../clients/task/taskClient'

const importanceOptions = ['High', 'Medium', 'Low']

export const TaskDisplay = ({ id, name, deadline, initialImportance, initialIsCompleted }) => {
  const [importance, setImportance] = useState(initialImportance)
  const [isCompleted, setIsCompleted] = useState(initialIsCompleted)

  const handleImportanceChange = (e) => {
    const newImportance = e.target.value
    setImportance(newImportance)
    taskClient.updateTask(id, { importance: newImportance })
  }

  const handleCompletionToggle = () => {
    const newIsCompleted = !isCompleted
    setIsCompleted(newIsCompleted)
    taskClient.updateTask(id, { isCompleted: newIsCompleted })
  }

  return (
    <tr className={`${styles.taskDisplay} ${isCompleted ? styles.completed : ''}`}>
      <ListCell>{name}</ListCell>
      <ListCell>{deadline}</ListCell>
      <ListCell>
        <ListSelect
          value={importance}
          options={importanceOptions}
          se
          onChange={handleImportanceChange}
        />
      </ListCell>
      <ListCell>
        <span style={{ width: '100%', display: 'flex', justifyContent: 'space-between' }}>
          <span style={{ width: '30px', display: 'inline-block' }}>
            {isCompleted ? 'Yes' : 'No'}
          </span>
          <ListCheckbox isToggled={isCompleted} onChange={handleCompletionToggle} />
        </span>
      </ListCell>
    </tr>
  )
}

TaskDisplay.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  deadline: PropTypes.string.isRequired,
  initialImportance: PropTypes.oneOf(importanceOptions).isRequired,
  initialIsCompleted: PropTypes.bool.isRequired
}
