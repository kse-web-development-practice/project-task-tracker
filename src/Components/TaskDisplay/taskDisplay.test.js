import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import { TaskDisplay } from './taskDisplay'
import taskClient from '../../clients/task/taskClient'

jest.mock('../../clients/task/taskClient', () => ({
  updateTask: jest.fn()
}))

describe('TaskDisplay Component', () => {
  const renderTaskDisplay = () =>
    render(
      <table>
        <tbody>
          <TaskDisplay
            id="1"
            name="Task 1"
            deadline="2024-12-31"
            initialImportance="Medium"
            initialIsCompleted={false}
          />
        </tbody>
      </table>
    )

  test('snapshot', () => {
    const { asFragment } = renderTaskDisplay()
    expect(asFragment()).toMatchSnapshot()
  })

  test('importance change', () => {
    renderTaskDisplay()

    fireEvent.change(screen.getByRole('combobox'), { target: { value: 'High' } })
    expect(taskClient.updateTask).toHaveBeenCalledWith('1', { importance: 'High' })
  })

  test('isCompleted change', () => {
    renderTaskDisplay()

    const checkbox = screen.getByRole('checkbox')
    fireEvent.click(checkbox)
    expect(taskClient.updateTask).toHaveBeenCalledWith('1', { isCompleted: true })
  })
})
