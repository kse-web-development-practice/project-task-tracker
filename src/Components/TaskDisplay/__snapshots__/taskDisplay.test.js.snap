// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskDisplay Component snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr
        class="taskDisplay "
      >
        <td
          class="listCell"
        >
          Task 1
        </td>
        <td
          class="listCell"
        >
          2024-12-31
        </td>
        <td
          class="listCell"
        >
          <select
            class="listSelect"
          >
            <option
              value="High"
            >
              High
            </option>
            <option
              value="Medium"
            >
              Medium
            </option>
            <option
              value="Low"
            >
              Low
            </option>
          </select>
        </td>
        <td
          class="listCell"
        >
          <span
            style="width: 100%; display: flex; justify-content: space-between;"
          >
            <span
              style="width: 30px; display: inline-block;"
            >
              No
            </span>
            <input
              class="listCheckbox"
              type="checkbox"
            />
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
